### DISCUSSION TITLE
User C submodule?

### DISCUSSION THREAD

**Discussion by MHe999** [micropython/micropython#14171] [Libraries & Drivers] [answered]
Hello!

We have currently a well-working set of custom C modules. They are imported and used in the Python code like this:

```
import module_x

module_x.function_y()
```

To avoid any name conflicts between our and MicroPython builtin modules (current and future), we would like to import our modules as submodules like this:

`from our_package import module_x`

After imported like this, "module_x" should be usable like before, for example: 

`module_x.function_y()`

and not like this: 

`our_package.module_x.function_y()`

Also, if possible, the modules such as "module_x" should be imported only when explicitly listed like above, meaning that if our_package is imported like this: 

`import our_package`

it shouldn't import all of the subpackages.

Is it possible to use subpackages with user C modules, and are they the correct way to achieve the described behavior?

Thanks in advance for any advice.

**Comment by dlech** [comment]
I think the only way to do this right now is to override `mp_builtin___import___default` by defining `mp_builtin___import__` and implementing your own module name resolution. There are also some older pull requests discussion various approaches to how this could be done for C modules in general, IIRC.

**Comment by jimmo** [comment, marked as answer]
@MHe999 You can implement built-in "sub" modules -- you need to enable `MICROPY_MODULE_BUILTIN_SUBPACKAGES`

See https://github.com/micropython/micropython/blob/master/py/mpconfig.h#L894 and https://github.com/micropython/micropython/tree/master/examples/usercmodule/subpackage

**Comment by MHe999** [comment]
Thanks @dlech and @jimmo for your replies.

Jimmo's way seems to do just what I wanted, perfect!
